### DISCUSSION TITLE
Need help  with filter in jinja2  in export template.

### DISCUSSION THREAD

**Discussion by nikunj-tele** [netbox-community/netbox#15208] [Help Wanted!] [answered]
Hi ,

I am getting following output on rack position

>>> d = Device.objects.get(name="device01")
>>> d.position
Decimal('20.0')

I only want to filter out  '20' from the above output when in export template.

tried to find  jinja2 [built in filter](https://jinja.palletsprojects.com/en/2.11.x/templates/#list-of-builtin-filters) but could not find the appropritate output 
I may be able to sort out the issue in nbshell as it use the python string but how can i get rack position output using jinja2 (export template)

**Comment by ross-cello** [comment]
Hm, can you do `int(dv.position)` ? I'm guessing not

Previously I used rsplit in my jinja2 template to achieve my goal of intepreting a subif as a VLAN value.
```
{% for ifd1q in device.interfaces.filter(name__contains=".") %}
{% set dot1q = ifd1q.name.rsplit('.', 1) %}
!
interface {{ ifd1q.name }}
 encapsulation dot1q {{ dot1q[1] }}
{% endfor %}
!
```
which gave me
```
!
interface GigabitEthernet0/0/0.10
 encapsulation dot1q 10
!
interface GigabitEthernet0/0/1.20
 encapsulation dot1q 20
!
```

**Reply by nikunj-tele** [reply]
Hi @ross-cello ,
Thank you for your input,
rspilt did not work 
it gave error 

**"decimal.Decimal object' has no attribute 'rspilt'"** 

my export template as below
```
{%- for device in queryset %}
{% set t1 = device.tags.slugs() | join("") -%}
{% set p1 = device.position.rspilt('.',1) %}
{% if t1 == 'test' %}
{{device.site}},{{device.rack.name}},{{ p1 }},{{device.device_type}},{{device.name}},{{ p1 }},{{device.serial}}
{% endif %}
{% endfor -%}

**Comment by kyerlasswell** [comment]
The `//` operator truncates a decimal. You can get the position number without the decimal by simply dividing the position by 1 with the `//` operator. 

```
d.position // 1

Decimal('20')
```
*edit: removed section about normalize() because it doesn't do what I thought.*

**Reply by nikunj-tele** [reply]
Thanks @kyerlasswell  for your answer.

I want almost same out put but still I am getting 

>>> d.position.normalize() // 1
Decimal('20')
so in this case I dont want the "Decimal" only '20' is required.
Additionaly, I also want to use with jina2 string in export template

**Reply by kyerlasswell** [reply]
So the *Decimal('20')* notation is simply netbox/nbshell being nice and telling you the type of object it's returning, but the actual returned value is just 20. When referenced somewhere, like in your template, it should give you the number alone, and you can check this for yourself by doing something like this in nbshell

```
print(d.position // 1)
20
```

**Reply by nikunj-tele** [reply]
Thanks @kyerlasswell 
Yes, I am testing it on nbshell 
>>> print(d.position // 1)
20
>>>
but seems same pattern doesnot accepted by jinja2 in export template like (print , // ) are not acceptable,  yes jinja2 is giving only number but it is incorret one  as you can see below output jinja2 gives '200' and the actual value is 20 so struggling with filters in export template
>>> d.position
Decimal('20.0')
>>>

**Reply by kyerlasswell** [reply, marked as answer]
I was just saying python's `print()` gives you an example of the returned data, you shouldn't use that in the template.

I just set up a simple test using export templates and I am having no problem with getting the device position in the format you're looking for. Again, the output that you get from nbshell, `Decimal('20')`, is just how the data is displayed - the actual returned data is the number alone.

My entire code for this test is exactly as shown below. I would recommend setting up an identical test for yourself and using that as a starting place for troubleshooting.
* The template loops over a queryset of any device list you want to export. 
* `d` is equal to the current device in the loop.
* The if statement skips devices with no position (otherwise it will error if it encounters a device with no position).

Export template:
```
{%- for d in queryset %}
{%- if d.position %}
Device position: {{ d.position // 1 }}
{%- endif %}
{%- endfor %}
```

Output file:
```
Device position: 25
Device position: 40
Device position: 37
```

*I'm also just now realizing `.normlize()` will strip the zeros off of more than just after the decimal. `20.0` becomes `2E+1`, so it shouldn't be used here like I thought. I wasn't testing on a device with that specific position, so that's my bad. I will edit my original comment.*

**Reply by nikunj-tele** [reply]
Thanks @kyerlasswell ,

it works well with if statement. (but could not understand why to use if statment ) do you think it is  jinja2 bug?

**Reply by kyerlasswell** [reply]
Glad I could help you find a solution. 

Nope, it's no bug, it's just that without the if statement, you are performing an invalid operation on devices that do not have a position (i.e. are not mounted in the rack). 

If a device is not mounted in a rack, its `position` is actually `None` instead of a number. Since you cannot divide `None // 1`, it throus this error:

```
unsupported operand type(s) for //: 'NoneType' and 'int' 
```

This is telling you that can't use the `//` operator with `None` and `1` because they are two different data types (NoneType and integer).

Adding the if statement checks if the `position` field contains any data. If it contains no data (`None`), that device is skipped and no operations are performed on it.

**Reply by nikunj-tele** [reply]
Thanks again @kyerlasswell ,

I got trouble with other field but here does not contain int value

in racklocaiton output I have following output again // 1 is not working as it it is non int value.
so how can i filter data like - M1  only and 1010 like a way in jinja2
```
>>> print(d.location)
M1.2.1010
>>> print(d.location // 1)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: unsupported operand type(s) for //: 'Location' and 'int'
>>>

**Reply by kyerlasswell** [reply]
You only need the `// 1` on the rack position because it gets rid of the decimal that you don't want. You can just do `d.location` to get the location object or `d.location.name` to get the location name as a string.

nbshell:
```
d.location.name
'Location Name'
```

export template (I haven't tested this):
```
{{ d.location.name }}
```
output:
```
Location Name
```

**Comment by nikunj-tele** [comment]
Hi @kyerlasswell ,

I have just tested on both nbshell and in jinja2

```
>>> d.location.name
'M1.2.1010'
>>>
```
jinja2 also gave  same output "'M1.2.1010'"

**Reply by kyerlasswell** [reply]
As far as that goes, I think I will have to leave you on your own, I can't really dedicate any more time to this and the solution is going to heavily depend on your location naming convention and making a decision on how to spit the name up using filters is going to take some work to make sure it is reliable across all possible location names.

As a parting suggestion to maybe point you in the right direction for your research, it looks like your location naming is hierarchical and you are looking to grab a specific part of that hierarchy. If I'm wrong ignore this, but it is probably worth looking into nesting your locations so that `1010` is its own location and is a child of `2` which is its own location and is a child of `M1`. You can use the parent field when you create locations to set up this hierarchy. Your locations would look like this (I'm filling in the blanks since I don't know your locations):

```
M1
  - 1
    - 1000
    - 2000
  - 2
    - 1010
M2
  - 1
    - 1234
...
etc
```
From here, you could query the exact location and any of its parents you want without having to rely on potentially complex filters.

Good luck to you.

**Reply by nikunj-tele** [reply]
Thanks @kyerlasswell , sorry I was in training, I will check this filter and also work on netbox to simply this output.
